Sources/ComposableArchitecture/Instrumentation.swift

```diff
@@ -0,0 +1,142 @@
+
+/// Interface to enable tracking/instrumenting the activity within TCA as ``Actions`` are sent into ``Store``s and
+/// ``ViewStores``, ``Reducers`` are executed, and ``Effects`` are observed.
+///
+/// The way the library will call the closures provided is identical to the way that the ``Actions`` and ``Effects`` are
+/// handled internally. That means that there is likely to be ``Instrumentation.ViewStore`` `will|did` pairs contained
+/// within the bounds of an ``Instrumentation.Store`` `will|did` pair. For example: Consider sending a simple ``Action``
+/// into a ``ViewStore`` that does not produce any synchronous ``Effects`` to be generated by a ``Reducer``, and the
+/// ``ViewStore`` is scoped off a parent ``Store``s state:
+///   1. An ``Action`` is passed to ``ViewStore.send``
+///   2. The ``Instrumentation.ViewStore.willSend`` callback is called
+///   3. The ``Action`` is passed to ``Store.send``
+///   4. The ``Instrumentation.Store.willSend`` callback is called
+///   5. The ``Instrumentation.Store.willProcessEvents`` callback is called
+///   6. The ``Action`` is passed to the ``Store.reducer`` and acted upon
+///   7. The ``Store.reducer`` returns an ``Effect.none`` producing no further ``Action``s
+///   8. The ``Instrumentation.Store.didProcessEvents`` callback is called
+///   9. The ``Instrumentation.Store.willScope`` callback is called
+///   10. The ``Store.state.value`` is updated
+///   11. The ``ViewStore`` is updated with the ``Store.state``s new value
+///   12. The ``Instrumentation.ViewStore.willDeduplicate`` callback is called
+///   13. The state is deduplicated
+///   14. The ``Instrumentation.ViewStore.didDeduplicate`` callback is called
+///   15. The ``Instrumentation.ViewStore.stateWillChange`` callback is called
+///   16. The ``ViewStore.state`` value is updated
+///   17. All subscribers to the ``ViewStore.state`` are updated with the new state value
+///   18. The ``Instrumentation.ViewStore.stateDidChange`` callback is called
+///   19. The ``Instrumentation.Store.didScope`` callback is called
+///   20. The ``Instrumentation.Store.didSend`` callback is called
+///   21. The ``Instrumentation.ViewStore.didSend`` callback is called
+public class Instrumentation {
+  public typealias Trigger = (EventInfo) -> Void
+
+  /// Container for the information that will be provided to tracking/instrumentation implementations.
+  public struct EventInfo: CustomStringConvertible {
+    internal init(type: String, action: String = "", tags: [String: String] = [:]) {
+      self.type = type
+      self.action = action
+      self.tags = tags
+    }
+
+    /// The Swift type of object that is operating. This will generally be of the type `Store<State, Action>` or
+    /// `ViewStore<State, Action>` with the `State` and `Action` types properly filled in. With this information it
+    /// _should_ be possible to identify which ``Store`` or ``ViewStore`` (or other type) is operating.
+    public let type: String
```

Comment:
A bit surprised this traffics in Strings (type & action); sorta assumed EventInfo would be generic over them.

Reply:
I will have to test it out. I think leaking the store and action types out to the implementation will prevent a global instance.

Reply:
Took a stab at passing the action and store's type instead of String values. Is pretty nice at the call site. But it ends up requiring type erasure (to `Any`) for both to prevent the generics from leaking out into the implementation of the callbacks. The follow on impact of that _seems_ to be that generating the store's type string is slower, but I've only done light profiling of this and so am not absolutely certain. Would definitely enjoy puzzling through this with you!